File: Submissions/662_Maximum_Width_of_Binary_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int widthOfBinaryTree(TreeNode* root) {
        queue<pair<TreeNode*, int> > q;
        q.push(make_pair(root, 1));
        int width = 1;
        while(!q.empty()){
            int n = q.size();
            if(n == 1) q.front().second = 1;
            width = max(width, q.back().second-q.front().second+1);
            for(int i=0; i<n; ++i){
                TreeNode* node = q.front().first;
                int index = q.front().second;
                q.pop();
                if(node->left) q.push(make_pair(node->left, index*2-1));
                if(node->right) q.push(make_pair(node->right, index*2));
            }
        }
        return width;
    }
};

File: Kotlin/662_Maximum_Width_of_Binary_Tree.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun widthOfBinaryTree(root: TreeNode?): Int {
        if (root == null) return 0

        var width = 1
        val q = LinkedList<Pair<TreeNode, Int>>()
        q.offer(Pair(root, 1))

        while(!q.isEmpty()){
            val n = q.size
            width = maxOf(width, q.last.second - q.first.second + 1)
            for(i in 0 until n) {
                val node = q.first.first
                val index = q.first.second
                q.pop()
                if(node.left != null) q.offer(Pair(node.left, index * 2 - 1))
                if(node.right != null) q.offer(Pair(node.right, index * 2))
            }
        }
        return width
    }
}